language/stdlib/modules/designated_dealer.move

```diff
@@ -0,0 +1,150 @@
+address 0x0 {
+    module DesignatedDealer {
+        use 0x0::LibraTimestamp;
+        use 0x0::Vector;
+        use 0x0::Transaction as Txn;
+        use 0x0::Libra;
+        use 0x0::LibraAccount;
+        use 0x0::Coin1;
+        use 0x0::Coin2;
+
+        resource struct Dealer {
+            // Time window start in microseconds
+            window_start: u64,
+            // The minted inflow during this time window
+            window_inflow: u64,
+            // The burned outflow during this time window
+            window_outflow: u64,
+            // Association grants
+            is_certified: bool,
+        }
+
+        resource struct Tiers {
+            // Tier-index to start of tier
+            tiers: vector<u64>
+        }
+
+        public fun initialize() {
+            assert_sender_is_blessed();
+            move_to_sender(Tiers {
+                tiers: Vector::empty(),
+            })
+        }
+
+        public fun add_tier(next_tier_upperbound: u64)
+        acquires Tiers {
+            assert_sender_is_blessed();
+            let tiers = borrow_global_mut<Tiers>(singleton_addr());
+            let number_of_tiers: u64 = Vector::length(&tiers.tiers);
+            // INVALID TIER_ADDITION
+            Txn::assert(number_of_tiers <= 4, 3);
+            if (number_of_tiers > 1) {
+                let prev_tier = *Vector::borrow(&tiers.tiers, number_of_tiers - 1);
+                // INVALID_TIER_START
+                Txn::assert(prev_tier < next_tier_upperbound, 4);
+            };
+            Vector::push_back(&mut tiers.tiers, next_tier_upperbound);
+        }
+
+        public fun update_tier(tier_index: u64, new_upperbound: u64)
+        acquires Tiers {
+            assert_sender_is_blessed();
+            let tiers = borrow_global_mut<Tiers>(singleton_addr());
+            let number_of_tiers = Vector::length(&tiers.tiers);
+            // INVALID_TIER_INDEX
+            Txn::assert(tier_index <= 4, 3);
+            Txn::assert(tier_index >= 0, 3);
+            Txn::assert(tier_index < number_of_tiers, 3);
+            // Make sure that this new start for the tier is consistent
+            // with the tiers above it.
+            let index = tier_index + 1;
+            while (index < number_of_tiers) {
+                // INVALID_TIER_START
+                Txn::assert(new_upperbound < *Vector::borrow(&tiers.tiers, index), 4);
+                index = index + 1;
+            };
+            let tier_mut = Vector::borrow_mut(&mut tiers.tiers, tier_index);
+            *tier_mut = new_upperbound;
+        }
+
+        public fun apply_for_designated_dealer() {
```

Comment:
Now that https://github.com/libra/libra/commit/e66e1fc5c9d61a5aace906ec01b60429beb6b485 is in, we could consider making a separate `Account::Role` for `DesignatedDealer`. I think this might make sense because otherwise we will have to pick one of the existing `Account::Role`s for DD and nothing that exists quite fits (VASP, association, unhosted)... In addition,  that would let us do a one-shot creation instead of using the apply/certify  pattern.

Reply:
Is the plan to standardize all accounts to LibraAccount::T with some specific RoleData?

Reply:
> Is the plan to standardize all accounts to LibraAccount::T with some specific RoleData?

Yes, that is the idea.

> Also, this introduces a cyclic dependency since mint_to_designated_dealer calls LibraAccount::deposit. Any suggestions of how to proceed that work with the current architecture?

I would address this by implementing `mint_to_designated_dealer` inside `LibraAccount` instead. In fact, I think you can just repurpose the existing `mint` to (1) check for the `DesignatedDealer` role (2) extract the DD `RoleData` and do the tier checks.